fs/api_internal.go

```diff
@@ -1515,319 +1515,757 @@ Restart:
+			if markerPathDirInodeIndex == (len(markerPath) - 1) {
+				prevReturned = ""
+			} else {
+				prevReturned = markerPath[len(markerPath)-1]
 			}
+			initialDirEntryToMatch = ""
+			break
+		}
 
-			if fileName > prefix && !strings.HasPrefix(fileName, prefix) {
-				// Remember that we're going over these in order, so the first time we see something that's greater that
-				// the prefix but doesn't start with it, we can skip the entire rest of the directory entries since they
-				// are *also* greater than the prefix but don't start with it.
-				return nil
+		switch strings.Compare(prefixPath[pathIndex], markerPath[pathIndex]) {
+		case -1:
+			dirPath = strings.Join(markerPath[:markerPathDirInodeIndex+1], "/")
+			if markerPathDirInodeIndex == (len(markerPath) - 1) {
+				prevReturned = ""
+			} else {
+				prevReturned = markerPath[len(markerPath)-1]
 			}
-			if endmarker != "" && fileName >= endmarker {
-				// Similarly, when we encounter something past our endmarker, we can skip the rest of the directory entries
-				// since they are *also* greater than the endmarker
-				return nil
+			initialDirEntryToMatch = ""
+			break
+		case 0:
+			pathIndex++
+		case 1:
+			if prefixPathDirInodeIndex == (len(prefixPath) - 1) {
+				if (1 == len(prefixPath)) || strings.HasSuffix(prefix, "/") {
+					dirPath = strings.Join(prefixPath[:prefixPathDirInodeIndex+1], "/")
+					prevReturned = ""
+				} else {
+					dirPath = strings.Join(prefixPath[:prefixPathDirInodeIndex], "/")
+					prevReturned = prefixPath[len(prefixPath)-1]
+				}
+			} else {
+				dirPath = strings.Join(prefixPath[:prefixPathDirInodeIndex+1], "/")
+				prevReturned = prefixPath[len(prefixPath)-1]
 			}
+			initialDirEntryToMatch = prevReturned
+			break
+		}
+	}
 
-			// Swift container listings are paginated; you
-			// retrieve the first page with a simple GET
-			// <container>, then you retrieve each subsequent page
-			// with a GET <container>?marker=<last-obj-returned>.
-			//
-			// If we were given a marker, then we can prune the
-			// directory tree that we're walking.
-			//
-			// For a regular file, if its container-relative path
-			// is lexicographically less than or equal to the
-			// marker, we skip it.
-			//
-			// For a directory, if its container-relative path is
-			// lexicographically less than or equal to the marker
-			// and the marker does not begin with the directory's
-			// path, we skip it.
-			//
-			// Since no regular file's container-relative path
-			// starts with another regular file's
-			// container-relative path, we can make the following
-			// test prior to any Getstat() calls, avoiding
-			// unneeded IO.
-			if fileName <= marker && strings.Index(marker, fileName) != 0 {
-				continue
+	// Setup shortcuts/contants
+
+	dlmCallerID = dlm.GenerateCallerID()
+	inodeVolumeHandle = mS.volStruct.inodeVolumeHandle
+
+	// Compute initial response
+
+	restartBackoff = time.Duration(0)
+
+Restart:
```

Comment:
shouldn't labels be lower case?

Reply:
Interesting... didn't know there was a convention.
Indeed, I was just trying to follow the pattern of other label gotos both in this file and elsewhere.
Happy to adopt your preferred naming convention.
There are probably only a handful of other files that actually use goto's :-)